stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -833,6 +833,281 @@ opnorm(v::AdjointAbsVec, q::Real) = q == Inf ? norm(conj(v.parent), 1) : norm(co
+            # replace ind[1:t] by first t indices in ind not in ind_hist
+            j = 1
+            for i in 1:n
+                indi = ind[i]
+                if indi ∉ ind_hist
+                    ind[j] = indi
+                    j += 1
+                end
+                j > t && break
+            end
+        end
+        for j = 1:t
+            fill!(view(X,1:n,j), 0)
+            X[ind[j],j] = 1
+        end
+        union!(ind_hist, view(ind,1:t))
+    end
+    ret = est
+    # (6)
+    if retv
+        v = zeros(Ti, n)
+        v[ind_best] = 1
+        ret = (ret, v)
+    end
+    if retw
+        ret = (ret..., w)
+    end
+    return ret
+end
+
+# Linear operator representing product of matrices A₁ * A₂ * …
+# utility type for opnormestprod1
+struct ProductMat{T}
+    As::T
+end
+Base.eltype(M::ProductMat) = Base.promote_eltype(M.As...)
+Base.size(M::ProductMat) = (size(first(M.As))[1], size(last(M.As))[2])
+function Base.:*(M::ProductMat, X)
+    for A in reverse(M.As)
+        X = A * X
+    end
+    return X
+end
+function Base.adjoint(M::ProductMat)
+    Ast = map(adjoint, reverse(M.As))
+    return ProductMat(Ast)
+end
+
+"""
+    opnormestprod1(As...)
+
+Estimate the operator 1-norm [`opnorm(prod(As), 1)`](@ref) of a product of matrices or
+linear operators without forming the product.
+
+Each element `A` in `As` can be of any type `Op`, representing a matrix or vector, that
+implements the following methods:
+- `size(A::Op)`
+- `eltype(A::Op)`
+- `*(A::Op, B::AbstractMatrix)`
+- `adjoint(A::Op)`
+Additionally, the first dimension of `As[1]` and the last dimension of `As[end]` must be
+equal for their product to be square.
+
+See [`opnormest1`](@ref) for details on the algorithm that computes the estimate.
+
+    opnormestprod1(A, p::Integer)
+
+Estimate the operator 1-norm for the matrix power `A^p` without exponentiating the matrix.
+"""
+opnormestprod1
+
+function opnormestprod1(A1, A2, As...)
+    return opnormest1(ProductMat([A1, A2, As...]))
+end
+function opnormestprod1(A, p::Integer)
+    return opnormest1(ProductMat(repeat([A], p)))
+end
+
+# Linear operator representing inverse of a matrix
+# utility type for opnormestinv
+struct InvMat{T}
+    A::T
+end
+Base.eltype(M::InvMat) = eltype(M.A)
+Base.size(M::InvMat) = reverse(size(M.A))
+Base.:*(M::InvMat, X) = M.A \ X
+Base.adjoint(M::InvMat) = InvMat(adjoint(M.A))
+
+"""
+    opnormestinv1(A, args...)
+
+Estimate the operator 1-norm [`opnorm(inv(A), 1)`](@ref) of a matrix or linear operator
+`A` without computing the inverse.
+
+See [`opnormest1`](@ref) for a description of the arguments and return values.
+
+`A` can be of any type `Op`, representing a square matrix, that implements the following
+methods:
+- `size(A::Op)`
+- `eltype(A::Op)`
+- `\\(A::Op, B::AbstractMatrix)`
+- `adjoint(A::Op)`
+
+If `A` is an `AbstractMatrix`, it will be more efficient to pass its factorization
+to this function.
+"""
+function opnormestinv1(A, args...)
+    checksquare(A)
```

Comment:
As written this currently also works to estimate `opnorm(pinv(A), 1)` for nonsquare `A`, except for this check, added for consistency with SparseArray's older version.

Reply:
Refactored to support `pinv`